### PR TITLE
Add struct type support for `drop_list_duplicates`

### LINKED ISSUE #8972: [FEA] add nested struct support for collect_set and merge_set aggregations
**Is your feature request related to a problem? Please describe.**
In the Spark plugin we are pushing to try and support structs more fully.  collect_list and merge_list already support structs, but collect_set and merge_set do not because the deduplication code cannot tell if nested values are the same.  Cudf already support sorting structs of structs so it should be able to come up with a similar way to tell if these are equal and deduplicate them.  To be 100% clear this is only for structs of basic types and structs of structs.  We done need list support for this, because that appears to be a lot more difficult.

### PR DESCRIPTION
**Pull request by ttnghia**
This PR add support for struct type into the existing `drop_list_duplicates` API. This is the first time a nested type is supported in this function. Some more code cleanup has also been done.

To be clear: Only structs of basic types and structs of structs are supported. Structs of lists are not, due to their complex nature.

Closes #8972.
Blocked by https://github.com/rapidsai/cudf/pull/9218 (it is merged).

### CODE REVIEW

File: cpp/include/cudf/detail/gather.cuh [hunk 1]
```diff
@@ -450,16 +450,24 @@ struct column_gatherer_impl<struct_view> {
                                      rmm::cuda_stream_view stream,
                                      rmm::mr::device_memory_resource* mr)
   {
-    structs_column_view structs_column(column);
-    auto gather_map_size{std::distance(gather_map_begin, gather_map_end)};
+    auto const gather_map_size = std::distance(gather_map_begin, gather_map_end);
```
Comment (ttnghia) [inline cpp/include/cudf/detail/gather.cuh:453]: Changes in this `gather.cuh` file is not part of this PR. Instead, it is upstream from another PR (https://github.com/rapidsai/cudf/pull/9218). Please don't review it here.

**Comment by vyasr** [comment]
@ttnghia it looks like changing the default operator overload for `has_negative_nans_dispatch` to CUDF_FAIL instead of returning false breaks APIs that were relying on that behavior. Is there an important reason to make that change?

**Comment by ttnghia** [comment]
> @ttnghia it looks like changing the default operator overload for `has_negative_nans_dispatch` to CUDF_FAIL instead of returning false breaks APIs that were relying on that behavior. Is there an important reason to make that change?

Oh my bad---the result of copy-then-edit mistake. Thanks so much for catching this bug. I have fixed it (change the code back to previous state).

File: cpp/src/lists/drop_list_duplicates.cu [hunk 2]
```diff
@@ -462,22 +572,23 @@ std::unique_ptr<column> drop_list_duplicates(lists_column_view const& lists_colu
...
   auto const sorted_lists = [&]() {
     // If nans_equal == ALL_EQUAL and the column contains lists of floating-point data type,
-    // we need to replace -NaN by NaN before sorting
+    // we need to replace -NaN by NaN before sorting.
     auto const replace_negative_nan =
-      nans_equal == nan_equality::ALL_EQUAL and
-      type_dispatcher(lists_entries.type(), detail::has_negative_nans_fn{}, lists_entries, stream);
+      nans_equal == nan_equality::ALL_EQUAL &&
+      type_dispatcher(
+        lists_entries.type(), detail::has_negative_nans_dispatch{}, lists_entries, stream);
```
Comment (jrhemstad) [inline cpp/src/lists/drop_list_duplicates.cu:590]: It seems redundant to first check if negative nan values exist before launching a kernel to replace negative nans. In the case where negative NaNs exist, it would be faster to just always run NaN replacement kernel. In the case where NaN's _don't_ exist, I believe unnecessarily running the NaN replacement function will be just as fast as checking if NaNs exist.

Comment (jrhemstad) [reply cpp/src/lists/drop_list_duplicates.cu:590]: Also, is replacing the NaNs necessary? In the regular `cudf::sort` we already treat all NaNs as equivalent in the row comparator, so there isn't a need there to do any NaN replacement.

Comment (ttnghia) [reply cpp/src/lists/drop_list_duplicates.cu:590]: Yeah, but there is one issue. If we call replace NaN, we will do copying---materialize a new column (which may be nested) and copy data from the old column over it. The result will always be a new column regardless of whether it has NaN or not. I try to avoid that by probing NaN before calling to replacing NaN.

Comment (ttnghia) [reply cpp/src/lists/drop_list_duplicates.cu:590]: For large columns (nested---structs columns with many members), I believe that avoiding unnecessary copy is beneficial.

Comment (jrhemstad) [reply cpp/src/lists/drop_list_duplicates.cu:590]: Oh, I was thinking we could do the replace NaN operation in place.

File: cpp/src/lists/drop_list_duplicates.cu [hunk 3]
```diff
@@ -50,89 +54,110 @@ struct has_negative_nans {
 };
 
 /**
- * @brief A structure to be used along with type_dispatcher to check if a
- * `column_view` has any negative NaN entry
+ * @brief A structure to be used along with type_dispatcher to check if a column has any
+ * negative NaN value.
  */
-struct has_negative_nans_fn {
+struct has_negative_nans_dispatch {
```
Comment (nvdbaranec) [inline cpp/src/lists/drop_list_duplicates.cu:60]: Isn't it possible to have  `List<float>` coming through here that we'd want to check?

File: cpp/src/lists/drop_list_duplicates.cu [hunk 4]
```diff
@@ -168,7 +192,35 @@ std::unique_ptr<column> generate_clean_offsets(lists_column_view const& lists_co
...
+ * ends of the result column.
+ */
+std::unique_ptr<column> replace_negative_nans_entries(column_view const& lists_entries,
+                                                      lists_column_view const& lists_column,
+                                                      rmm::cuda_stream_view stream)
+{
+  // We need to copy the offsets column of the input lists_column. Since the input lists_column may
+  // be sliced, we need to generate clean offsets (i.e., offsets starting from zero).
+  auto new_offsets =
+    generate_clean_offsets(lists_column, stream, rmm::mr::get_current_device_resource());
```
Comment (nvdbaranec) [inline cpp/src/lists/drop_list_duplicates.cu:209]: Probably too late for this PR, but having a function that does this for lists (and strings) available as a utility seems like it would be useful.

File: cpp/src/lists/drop_list_duplicates.cu [hunk 5]
```diff
@@ -205,150 +256,211 @@ std::unique_ptr<column> generate_entry_list_offsets(size_type num_entries,
...
+  bool __device__ compare(T const& lhs_val, T const& rhs_val) const noexcept
   {
-    // Two entries are not considered for equality if they belong to different lists
+    // If both element(i) and element(j) are NaNs and nans are considered as equal value then this
+    // comparison will return `true`. This is the desired behavior in Pandas.
+    if (nans_equal && std::isnan(lhs_val) && std::isnan(rhs_val)) { return true; }
+
+    // If nans are considered as NOT equal, even both element(i) and element(j) are NaNs this
+    // comparison will still return `false`. This is the desired behavior in Apache Spark.
+    return lhs_val == rhs_val;
```
Comment (nvdbaranec) [inline cpp/src/lists/drop_list_duplicates.cu:289]: For the purposes of checking for duplicates, are we expecting to be checking for _exact_ duplicates or does potential floating point precision come into play?  If this is the latter, it might useful in the future to examine the overlap between this code and the equality/equivalency checking code in `cudf_test/column_utilities`.

**Review by nvdbaranec — CHANGES_REQUESTED** [review]

File: cpp/src/lists/drop_list_duplicates.cu [hunk 6]
```diff
@@ -50,89 +54,110 @@ struct has_negative_nans {
...
-                            thrust::make_counting_iterator(lists_entries.size()),
-                            detail::has_negative_nans<Type>{*d_entries, lists_entries.has_nulls()});
+    return thrust::count_if(
+      rmm::exec_policy(stream),
+      thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(lists_entries.size()),
+      detail::has_negative_nans_fn<Type>{*d_entries, lists_entries.has_nulls()});
+  }
+
+  template <typename Type, std::enable_if_t<std::is_same_v<Type, cudf::struct_view>>* = nullptr>
```
Comment (vyasr) [inline cpp/src/lists/drop_list_duplicates.cu:72]: I'm not entirely sure, is this really noexcept? There are a lot of function calls here (thrust and libcudf) and I'm not confident that all of them are marked that way or whether we might trigger UB here. Same goes for all the associated functors.

File: cpp/src/lists/drop_list_duplicates.cu [hunk 7]
```diff
@@ -205,150 +256,211 @@ std::unique_ptr<column> generate_entry_list_offsets(size_type num_entries,
...
 
-  template <bool nans_equal_ = nans_equal>
-  std::enable_if_t<cuda::std::is_floating_point_v<Type> and nans_equal_, bool> __device__
-  operator()(size_type i, size_type j) const noexcept
+  template <typename T, std::enable_if_t<cuda::std::is_floating_point_v<T>>* = nullptr>
+  bool __device__ compare(T const& lhs_val, T const& rhs_val) const noexcept
   {
-    // Two entries are not considered for equality if they belong to different lists
+    // If both element(i) and element(j) are NaNs and nans are considered as equal value then this
+    // comparison will return `true`. This is the desired behavior in Pandas.
```
Comment (vyasr) [inline cpp/src/lists/drop_list_duplicates.cu:284]: Maybe this is something we can improve in a future PR if it's too big a change now, but do we want to embed pandas/Spark specific behavior this deep in the call stack rather than having some configuration value passed down? I can see that it would need to get passed through about 4 levels of calls to get here, which would be annoying, but it also seems more in line with our general design philosophy to have an enum to configure this behavior rather than encoding it here.

File: cpp/src/lists/drop_list_duplicates.cu (hunk 3 above)
Comment (ttnghia) [reply cpp/src/lists/drop_list_duplicates.cu:60]: No. This functor is called on the child column of the input lists column. If the input is a `List<List<float>>` column, exception will be thrown so the child column `List<float>` will never be passed here.

**Review by vyasr — CHANGES_REQUESTED** [review]
A couple of minor questions, but this looks pretty close to ready.

File: cpp/src/lists/drop_list_duplicates.cu (hunk 4 above)
Comment (ttnghia) [reply cpp/src/lists/drop_list_duplicates.cu:209]: Yes, maybe very useful. I observed that this utility is also implemented in `lists/segmented_sort.cu`. I'll file an issue to add it in the future.

File: cpp/src/lists/drop_list_duplicates.cu (hunk 5 above)
Comment (ttnghia) [reply cpp/src/lists/drop_list_duplicates.cu:289]: This is an interesting question. I thought about this before when firstly implemented it, but since there is no requirement for adding precision control thus I just did an exact comparison. If there is a request for adding precision control then we can go back and add it easily.

Comment (jrhemstad) [reply cpp/src/lists/drop_list_duplicates.cu:289]: I don't think this is something we want to entertain. The implication would then be that _all_ libcudf functions that are comparing floating point values for equality should allow controlling the precision. This is just part of how floating point values work and while it makes sense to have that control in testing utilities, putting it in actual compute APIs would be very detrimental to maintenance/performance/testing.

File: cpp/src/lists/drop_list_duplicates.cu (hunk 6 above)
Comment (ttnghia) [reply cpp/src/lists/drop_list_duplicates.cu:72]: Thanks. Since this func involves recursive calls, it may throw. I fixed it.

File: cpp/src/lists/drop_list_duplicates.cu (hunk 7 above)
Comment (ttnghia) [reply cpp/src/lists/drop_list_duplicates.cu:284]: Absolutely agree with this. When this API was first reviewed, there was a long discussion about how to satisfy both Pandas' users and Spark's users, since each side requires a different behavior (NaN are equal or not). In the meantime, I don't have a better idea how to avoid this deep call stack. I'm happy to do it (or anybody can do it) if there is a better way.

Comment (vyasr) [reply cpp/src/lists/drop_list_duplicates.cu:284]: Maybe I'm missing something, but would it not be possible to define an enum for different nan behaviors and then pass that down (all the way through `get_unique_entries_and_list_offsets`->`get_unique_entries_dispatch`... I know it's a lot of forwarding) to control this behavior based on a templated version of the `column_row_comparator_fn`? I'm not saying that we need to do it in this PR, but if that solution would work and be preferable we should at least make an issue for this to avoid losing track once the PR is merged.

Comment (ttnghia) [reply cpp/src/lists/drop_list_duplicates.cu:284]: Thanks. I'll take a note on this. There should be a room for improvement.

File: cpp/src/lists/drop_list_duplicates.cu (hunk 4 above)
Comment (ttnghia) [reply cpp/src/lists/drop_list_duplicates.cu:209]: Done. Here it is: https://github.com/rapidsai/cudf/issues/9256

**Comment by vyasr** [comment]
rerun tests

**Review by vyasr — APPROVED** [review]

File: cpp/src/lists/drop_list_duplicates.cu (hunk 7 above)
Comment (ttnghia) [reply cpp/src/lists/drop_list_duplicates.cu:284]: Done. Here it is: https://github.com/rapidsai/cudf/issues/9257

**Comment by ttnghia** [comment]
Rerun tests.

**Review by harrism — CHANGES_REQUESTED** [review]
Minor comments inline.   For the record, I really dislike the -NaN replacement requirement. Ideally this should be factored out into an opt-in extension for platform-specific NaN behavior. But I guess libcudf is already riddled with platform-specific special casing at a very low level.

File: cpp/src/lists/drop_list_duplicates.cu [hunk 8]
```diff
@@ -50,89 +54,110 @@ struct has_negative_nans {
...
+      thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(lists_entries.size()),
+      detail::has_negative_nans_fn<Type>{*d_entries, lists_entries.has_nulls()});
+  }
+
+  template <typename Type, std::enable_if_t<std::is_same_v<Type, cudf::struct_view>>* = nullptr>
+  bool operator()(column_view const& lists_entries, rmm::cuda_stream_view stream) const
+  {
+    // Recursively check negative NaN on the children columns.
+    return std::any_of(
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/lists/drop_list_duplicates.cu:76]: Thanks Mark. Yes, I am aware of that thus issue but this is `std::`, not `thrust::` and it runs sequentially thus I don't think there is a problem with it.

[the remark this answers is not in the document]
Comment (harrism) [reply cpp/src/lists/drop_list_duplicates.cu:88]: Sorry, I misread.

**Review by harrism — APPROVED** [review]

**Comment by ttnghia** [comment]
@gpucibot merge
